ssh/generate.go

```diff
@@ -4,62 +4,123 @@
 	"encoding/pem"
 	"fmt"
+	mathrand "math/rand"
 	"strings"
 
 	"github.com/juju/errors"
 	"golang.org/x/crypto/ssh"
 )
 
-// rsaGenerateKey allows for tests to patch out rsa key generation
-var rsaGenerateKey = rsa.GenerateKey
+// PublicKey returns the public key for any private key. The public key is
+// suitable to be added into an authorized_keys file, and has the comment
+// passed in as the comment part of the key.
+func PublicKey(privateKey []byte, comment string) (string, error) {
+	signer, err := ssh.ParsePrivateKey(privateKey)
+	if err != nil {
+		return "", errors.Annotate(err, "failed to load key")
+	}
 
-// KeyBits is used to determine the number of bits to use for the RSA keys
-// created using the GenerateKey function.
-var KeyBits = 2048
+	authKey := string(ssh.MarshalAuthorizedKey(signer.PublicKey()))
+	// Strip off the trailing new line so we can add a comment.
+	authKey = strings.TrimSpace(authKey)
+	public := fmt.Sprintf("%s %s\n", authKey, comment)
+
+	return public, nil
+}
 
-// GenerateKey makes a 2048 bit RSA no-passphrase SSH capable key.  The bit
-// size is actually controlled by the KeyBits var. The private key returned is
-// encoded to ASCII using the PKCS1 encoding.  The public key is suitable to
-// be added into an authorized_keys file, and has the comment passed in as the
-// comment part of the key.
+// GenerateKey generated an Ed25519 no-passphrase SSH capable key. The private
+// key is returned in an OpenSSH compatible format. The public key is suitable
+// to be added into an authorized_keys file. The comment is encoded into the
+// private key, and passed into the comment part of the public key
 func GenerateKey(comment string) (private, public string, err error) {
-	key, err := rsaGenerateKey(rand.Reader, KeyBits)
+	pub, priv, err := ed25519.GenerateKey(rand.Reader)
 	if err != nil {
 		return "", "", errors.Trace(err)
 	}
 
+	publicKey, _ := ssh.NewPublicKey(pub)
+	public = string(ssh.MarshalAuthorizedKey(publicKey))
+	// insert comment to end of the public key in the comment space
+	public = fmt.Sprintf("%s %s\n", strings.TrimSpace(public), comment)
+
 	identity := pem.EncodeToMemory(
 		&pem.Block{
-			Type:  "RSA PRIVATE KEY",
-			Bytes: x509.MarshalPKCS1PrivateKey(key),
-		})
+			Type:  "OPENSSH PRIVATE KEY",
+			Bytes: MarshalEd25519PrivateKey(priv, pub, comment),
+		},
+	)
+	return string(identity), public, nil
+}
 
-	public, err = PublicKey(identity, comment)
-	if err != nil {
-		return "", "", errors.Trace(err)
+// MarshalEd25519PrivateKey formats an Ed25519 private key into a OpenSSH
+// compatible format
+//
+// NOTE(jack-w-shaw, 2023-01-20) Ideally we would a standard library function
+// to do this, as we can with RSA and ECDSA algorithms. But this doesn't seem
+// possible for Ed25519 as an exception. My implementation is based on
+// https://github.com/mikesmitty/edkey
+func MarshalEd25519PrivateKey(privateKey ed25519.PrivateKey, publicKey ed25519.PublicKey, comment string) []byte {
```

Comment:
OpenSSH supports PKCS8 since like v8.1. We don't need to write OpenSSH keys. See what I wrote in the `pki/ssh` package for generating ssh host_keys.